### PR TITLE
Add in weight based spool runout 

### LINKED ISSUE #638: [Enhancement]: Configurable Automatic Pre-End Spool Switching (Spoolman Threshold)
### What are you looking to see?

- Add a configurable automatic spool switch trigger based on remaining filament weight reported by Spoolman, designed primarily for Infinite Spool mode.

### Why do you want this feature?

**Problem**
When using Infinite Spool mode, the system waits for physical filament runout before switching spools. However, certain spools (notably some LDO spools) can develop a bent or hooked filament end near completion. This may:

- Prevent proper unload
- Cause filament to jam in PTFE paths
- Fail to trigger runout detection

Interrupt prints despite Infinite Spool being enabled

Because this is a mechanical end-of-spool issue, it is not always detected by filament sensors.

**Proposed Solution**
Introduce a configurable threshold (in grams) that automatically triggers a spool switch before the filament is fully depleted, based on remaining weight from Spoolman.

Proposed Configuration Options
```
[afc]
auto_spool_switch: False
auto_spool_switch_threshold: 25
auto_spool_switch_infinite_only: True
```

**Behavior:**
Feature disabled by default (fully backward compatible).

- When enabled
> - If remaining filament weight ≤ threshold
> - And Infinite Spool mode is active (optional restriction)
> - Automatically trigger spool change to next available spool

- Trigger should occur only once per spool (debounced).

**Intended Benefits**

- Prevents end-of-spool filament hook jams
- Reduces failed unload events
- Improves reliability of Infinite Spool setups
- No additional hardware required
- Fully optional feature

**Notes**

- Should gracefully do nothing if Spoolman is not connected.
- Should log when auto-switch is triggered.
- Should not interfere with existing manual or runout-based switching logic.

### Any other input or comments should go here.

_No response_

### PR DESCRIPTION
**Pull request by ejsears**
## Major Changes in this PR

This pull request introduces an automatic spool switch feature based on filament weight, along with supporting configuration options, logic, and tests. The main goal is to enable the system to automatically trigger a spool change when the remaining filament weight drops below a configurable threshold, improving print reliability and reducing manual intervention. The implementation includes configuration changes, logic updates in the main codebase, and comprehensive tests to ensure correct behavior and edge case handling.

**Auto Spool Switch Feature:**

* Added `auto_spool_switch` and `auto_spool_switch_threshold` options to `AFC.cfg` and integrated them into the configuration parsing and AFC initialization logic, allowing users to enable auto spool switching and set the weight threshold. [[1]](diffhunk://#diff-4d52fbf62af0688fc800a885ce481e17100e1723916a4bdbd0f8ac41ce10f0d8R37-R39) [[2]](diffhunk://#diff-749740ba904c224003143ed84a943eb5885a486b4ec55cd658a5beeea1cd1ff4R153-R156) [[3]](diffhunk://#diff-e52e4ddd58b7ef887ab03c04116e676f6280b824ab7469d5d3080e5cba4f2128R371-R372)
* Implemented weight-based auto spool switch logic in `AFC_lane.py`, including a debounce flag (`_auto_switch_triggered`) to prevent repeated triggers, and methods to handle the switch event by pausing or performing an infinite runout as appropriate. [[1]](diffhunk://#diff-120880cf9386cf18b6cdd101a1352383e1aa13474d51b2d99255ff11f286a279R126) [[2]](diffhunk://#diff-120880cf9386cf18b6cdd101a1352383e1aa13474d51b2d99255ff11f286a279R942-R954) [[3]](diffhunk://#diff-120880cf9386cf18b6cdd101a1352383e1aa13474d51b2d99255ff11f286a279R1311-R1326)
* Ensured the auto switch trigger flag is reset whenever spool values are set, cleared, or a new spool ID is assigned, preventing stale state across spool changes. [[1]](diffhunk://#diff-13e9a07c4eadfbd5a3e17927b682551beac5c0ce7d2ccf5315cff8bdb139985fR290) [[2]](diffhunk://#diff-13e9a07c4eadfbd5a3e17927b682551beac5c0ce7d2ccf5315cff8bdb139985fR305) [[3]](diffhunk://#diff-13e9a07c4eadfbd5a3e17927b682551beac5c0ce7d2ccf5315cff8bdb139985fR316)
* Fixed a typo in the `INVALID_UNIT_NAMES` list and improved section matching logic for unit configuration discovery. [[1]](diffhunk://#diff-120880cf9386cf18b6cdd101a1352383e1aa13474d51b2d99255ff11f286a279L44-R44) [[2]](diffhunk://#diff-120880cf9386cf18b6cdd101a1352383e1aa13474d51b2d99255ff11f286a279L664-R665)

**Testing:**

* Added comprehensive tests for the auto spool switch feature, covering correct triggering, debounce logic, logging, and edge cases such as error states and non-printing conditions. Also included tests to verify that the auto switch flag is properly reset during spool operations. [[1]](diffhunk://#diff-550044c902c6dceb6e3b5a8f92eb6c172507b14326249c4a61597dccff232fcaR828-R944) [[2]](diffhunk://#diff-38f66a43af34f26bc5348b36c9d78620e9fb1cefed03fc2b602285413be57e67R765-R808)

## Notes to Code Reviewers

Closes #638 

## How the changes in this PR are tested

Set infinite runout and appropriate options, and start print.

```
18:48:03   4194.312 [AFC_lane.py:update_weight_callback():1293]         - INFO: Auto spool switch: lane8 weight (649.9g) at or below threshold (650.0g), triggering switch
```
Verify it switches.

## PR Checklist: (Checked-off items are either done or do not apply to this PR)
 
- [x] I have performed a self-review of my code
- [x] CHANGELOG.md is updated (if end-user facing)
- [ ] Documentation updated in AT-Documentation repository
- [x] Sent notification to software-design/software-discussions channel (as appropriate) requesting review
- [x] If this PR address a GitHub issue, the issue number is referenced in the PR description

**NOTE**: GitHub Copilot may be used for automated code reviews, however as it is an automated system, it's suggestions 
may not be correct. Please do not rely on it to catch all issues. Please review any suggestions it makes and use your 
own judgement to determine if they are correct.

### CODE REVIEW

File: config/AFC.cfg [hunk 1]
```diff
@@ -34,6 +34,9 @@ assisted_unload: True           # If True, the unload retract is assisted to pre
 #print_short_stats: True          # Set to true to print AFC_STATS in short form instead of wide form, printing short form is better for consoles that are small in width
 #debug: True                     # Setting to True turns on more debugging to show on console
 
+#auto_spool_switch: False               # Trigger spool switch based on remaining filament weight from Spoolman
```
Comment (jimmyjon711) [inline config/AFC.cfg]: nit: Remove spoolman since this also applies to users that manually enter weight
```suggestion
#auto_spool_switch: False               # Trigger spool switch based on remaining filament weight
```

File: extras/AFC.py [hunk 2]
```diff
@@ -150,6 +150,10 @@ def __init__(self, config: ConfigWrapper):
 
         self.disable_weight_check   = config.getboolean("disable_weight_check", False) # Set to True to disable weight check when loading filament into lane/toolhead
         self.disable_ooze_check     = config.getboolean("disable_ooze_check", False) # Disable ooze check for lanes being on the same extruder in M104/M109 commands
+
+        # Auto spool switch settings
+        self.auto_spool_switch              = config.getboolean("auto_spool_switch", False)              # Trigger spool switch based on remaining filament weight
```
Comment (jimmyjon711) [inline extras/AFC.py]: Been trying to add typing to more things
```suggestion
        self.auto_spool_switch: bool          = config.getboolean("auto_spool_switch", False)              # Trigger spool switch based on remaining filament weight
```

File: extras/AFC.py [hunk 3]
```diff
@@ -150,6 +150,10 @@ def __init__(self, config: ConfigWrapper):
...
+        self.auto_spool_switch_threshold    = config.getfloat("auto_spool_switch_threshold", 25.0, minval=0.)  # Weight threshold in grams
```
Comment (jimmyjon711) [inline extras/AFC.py]:
```suggestion
        self.auto_spool_switch_threshold: float = config.getfloat("auto_spool_switch_threshold", 25.0, minval=0.)  # Weight threshold in grams
```

**Review by jimmyjon711 — CHANGES_REQUESTED** [review]
LGTM and tested with just pausing,

File: CHANGELOG.md [hunk 4]
```diff
@@ -5,9 +5,15 @@ All notable changes to this project will be documented in this file.
 The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.1.0/),
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
+<<<<<<< issue-638
```
Comment (jimmyjon711) [inline CHANGELOG.md]:
```suggestion
```

File: CHANGELOG.md [hunk 5]
```diff
@@ -5,9 +5,15 @@ All notable changes to this project will be documented in this file.
 The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.1.0/),
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
+<<<<<<< issue-638
+## [2026-03-23]
+### Added
+- Adds support for weight based spool runout to help support spools with hooked ends. See documentation for more details.
+=======
```
Comment (jimmyjon711) [inline CHANGELOG.md]:
```suggestion
```

File: CHANGELOG.md [hunk 6]
```diff
@@ -5,9 +5,15 @@ All notable changes to this project will be documented in this file.
...
 
+<<<<<<< issue-638
+## [2026-03-23]
+### Added
+- Adds support for weight based spool runout to help support spools with hooked ends. See documentation for more details.
+=======
 ## [2026-03-21]
 ### Fixed
 - Fixed initial load failure when an extruder is loaded with a different lane than the first in the print.
+>>>>>>> multi_extruder
```
Comment (jimmyjon711) [inline CHANGELOG.md]:
```suggestion
```

**Review by jimmyjon711 — CHANGES_REQUESTED** [review]
Remove, merge tags made it into the change log

**Review by jimmyjon711 — APPROVED** [review]
LGTM now
